### PR TITLE
[feat]: [CDS-58159]: changes to get overrides v2 in service step

### PR DESCRIPTION
**Pull request by tchaurasiya**
## Describe your changes

Pseudo Code Flow:
ServiceStep
   Facade -> getMergedServiceOverrideConfig(paramters)

Facade
getMergedServiceOverrideConfig(paramters) (returns Map<Type,NgOverrideConfig2> with runtime input merged)

	v1 -> 
		Env&Service|OverrideEntity - get().filter(yaml vs spec)
		mergeinputs()
		convertToNgOverrideConfig2() - config1
		map.add(envservicetype,config1)

		envEntity - get()
		megreinput()
		convertToNgOverrideConfig2() - config2
		map.add(envservicetype,config2)
		return map

	v2 --> 
		getAllOverride().filter(isv2True) - map<Location, list<entity>
		mergeScopeWise - map<Location, configv2>

		for - env-service - mergeinput  -  configV2
		for - env-global - merge input - configV2

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/48407)

### CODE REVIEW

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java [hunk 1]
```diff
@@ -74,15 +92,243 @@ public List<NGVariable> getOverrideVariables(@NonNull NGServiceOverridesEntity e
...
   }
 
-  public NGServiceOverrideConfigV2 mergeServiceOverridesInputs(
-      NGServiceOverridesEntity entity, Map<String, Object> serviceOverrideInputs) {
-    if (isEmpty(serviceOverrideInputs)) {
-      return getOverrideConfigV2ForNoInputs(entity);
+  public Map<String, NGServiceOverrideConfigV2> getMergedServiceOverrideConfigs(String accountId, String orgId,
+      String projectId, @NonNull ServiceStepV3Parameters parameters, @NonNull Environment envEntity)
+      throws IOException {
+    if (isEmpty(parameters.getEnvRef().getValue()) || isEmpty(parameters.getServiceRef().getValue())
```
Comment (yogesh-chauhan) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java:98]: can we check for ParamterField.isNotNull as well ?

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java [hunk 2]
```diff
@@ -74,15 +92,243 @@ public List<NGVariable> getOverrideVariables(@NonNull NGServiceOverridesEntity e
...
+        || isEmpty(parameters.getInfraId().getValue())) {
```
Comment (yogesh-chauhan) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java:99]: this should not throw error or log for now

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java [hunk 3]
```diff
@@ -74,15 +92,243 @@ public List<NGVariable> getOverrideVariables(@NonNull NGServiceOverridesEntity e
...
+      throw new InvalidRequestException("Environment Ref or Service Ref given for overrides has not been resolved");
     }
-    if (Boolean.TRUE.equals(entity.getIsV2())) {
-      return mergeOverrideV2Inputs(entity, serviceOverrideInputs);
+
+    if ((!ParameterField.isNull(parameters.getServiceOverrideInputs())
+            && parameters.getServiceOverrideInputs().isExpression())
+        || (!ParameterField.isNull(parameters.getEnvInputs()) && parameters.getEnvInputs().isExpression())) {
+      throw new InvalidRequestException("Service Override or Environment inputs are not resolved");
```
Comment (yogesh-chauhan) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java:106]: what is the value of this ? lets log it

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java [hunk 4]
```diff
@@ -74,15 +92,243 @@ public List<NGVariable> getOverrideVariables(@NonNull NGServiceOverridesEntity e
...
+    Map<String, NGServiceOverrideConfigV2> overridesMap = new HashMap<>();
+    String isOverrideV2EnabledValue =
+        NGRestUtils
+            .getResponse(ngSettingsClient.getSetting(OVERRIDE_PROJECT_SETTING_IDENTIFIER, accountId, orgId, projectId))
+            .getValue();
+
+    if (isOverrideV2EnabledValue.equals("true")) {
+      Map<String, List<NGServiceOverridesEntity>> allTypesOverridesV2 =
+          getAllOverridesWithSpecExists(parameters, accountId);
+      Map<String, NGServiceOverrideConfigV2> scopeWiseMergedOverrides =
```
Comment (yogesh-chauhan) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java:118]: nit: can we call this acrossScopes ?

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java [hunk 5]
```diff
@@ -74,15 +92,243 @@ public List<NGVariable> getOverrideVariables(@NonNull NGServiceOverridesEntity e
...
-      return mergeOverrideV2Inputs(entity, serviceOverrideInputs);
+
+    if ((!ParameterField.isNull(parameters.getServiceOverrideInputs())
+            && parameters.getServiceOverrideInputs().isExpression())
+        || (!ParameterField.isNull(parameters.getEnvInputs()) && parameters.getEnvInputs().isExpression())) {
+      throw new InvalidRequestException("Service Override or Environment inputs are not resolved");
     }
-    return mergeOverrideV1Inputs(entity, serviceOverrideInputs);
+
+    Map<String, NGServiceOverrideConfigV2> overridesMap = new HashMap<>();
```
Comment (yogesh-chauhan) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java:109]: can the key be enum TYPE ?

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java [hunk 6]
```diff
@@ -74,15 +92,243 @@ public List<NGVariable> getOverrideVariables(@NonNull NGServiceOverridesEntity e
...
+        Optional<NGServiceOverrideConfigV2> finalOverrideConfig = mergeOverridesGroupedByScope(entities);
+        finalOverrideConfig.ifPresent(
+            ngServiceOverrideConfigV2 -> finalMergedOverridesMap.put(type, ngServiceOverrideConfigV2));
+      }
+    });
+
+    return finalMergedOverridesMap;
+  }
+
+  private Optional<NGServiceOverrideConfigV2> mergeOverridesGroupedByScope(
```
Comment (yogesh-chauhan) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java:186]: nit: name can be grouped by type instead

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java [hunk 7]
```diff
@@ -74,15 +92,243 @@ public List<NGVariable> getOverrideVariables(@NonNull NGServiceOverridesEntity e
...
+      @NonNull List<NGServiceOverridesEntity> overridesEntities) {
+    if (overridesEntities.size() >= 3) {
+      throw new InvalidRequestException(
+          "Found more than 3 overrides, at one scope (project, org, account) only one override is supported");
```
Comment (yogesh-chauhan) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java:190]: this should be > 3 only

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java [hunk 8]
```diff
@@ -74,15 +92,243 @@ public List<NGVariable> getOverrideVariables(@NonNull NGServiceOverridesEntity e
...
+  }
+
+  private static void updateSpecFieldsByEntityFields(Map<String, NGVariable> finalNGVariables,
+      Map<String, ManifestConfigWrapper> finalManifests, Map<String, ConfigFileWrapper> finalConfigFiles,
+      ApplicationSettingsConfiguration finalApplicationSetting, ConnectionStringsConfiguration finalConnectionStrings,
+      ServiceOverridesSpec overridesSpec) {
+    overridesSpec.getVariables().forEach(ngVar -> finalNGVariables.put(ngVar.getName(), ngVar));
+    overridesSpec.getManifests().forEach(
+        manifest -> finalManifests.put(manifest.getManifest().getIdentifier(), manifest));
+    overridesSpec.getConfigFiles().forEach(
```
Comment (yogesh-chauhan) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java:259]: null check for override spec.getVariableS() and manifests/config files

**Review by yogesh-chauhan — CHANGES_REQUESTED** [review]

**Comment by tchaurasiya** [comment]
trigger uts

**Review by tchaurasiya — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48407)* status: 52 file(s) to review. 1 change requests. 6. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java` line 98 at r2](https://reviewable.io/reviews/harness/harness-core/48407#gh-1206692321:-NWOolBl5hgNNjIzoRFo:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/2fcf5a6b8da14708a0e7771c5b7a281665cfefa6/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java#L98)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

can we check for ParamterField.isNotNull as well ?
</blockquote></details>

Done.
___
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java` line 99 at r2](https://reviewable.io/reviews/harness/harness-core/48407#gh-1206698483:-NWOolbg0j3jl4lwecMI:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/2fcf5a6b8da14708a0e7771c5b7a281665cfefa6/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java#L99)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

this should not throw error or log for now
</blockquote></details>

Done.
___
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java` line 106 at r2](https://reviewable.io/reviews/harness/harness-core/48407#gh-1206699561:-NWOop3d4hpsZ8PzSHhx:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/2fcf5a6b8da14708a0e7771c5b7a281665cfefa6/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java#L106)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

what is the value of this ? lets log it
</blockquote></details>

Done.
___
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java` line 109 at r2](https://reviewable.io/reviews/harness/harness-core/48407#gh-1206703868:-NWOopa72j4xSfzRv41H:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/2fcf5a6b8da14708a0e7771c5b7a281665cfefa6/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java#L109)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

can the key be enum TYPE ?
</blockquote></details>

Done.
___
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java` line 190 at r2](https://reviewable.io/reviews/harness/harness-core/48407#gh-1206706305:-NWOos9A1njzHZCytiJc:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/2fcf5a6b8da14708a0e7771c5b7a281665cfefa6/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java#L190)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

this should be > 3 only
</blockquote></details>

Done.
___
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java` line 259 at r2](https://reviewable.io/reviews/harness/harness-core/48407#gh-1206710691:-NWOowAB0Z9pyM_Or6Ot:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/2fcf5a6b8da14708a0e7771c5b7a281665cfefa6/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java#L259)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

null check for override spec.getVariableS() and manifests/config files
</blockquote></details>

Done.

**Review by yogesh-chauhan — CHANGES_REQUESTED** [review]
> Reviewed 3 of 52 files at r1, 47 of 48 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48407)* status: 52 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java` line 99 at r2](https://reviewable.io/reviews/harness/harness-core/48407#gh-1206698483:-NWWA6oDB2jv-vptaxhX:b-f0exxg) ([raw file](https://github.com/harness/harness-core/blob/2fcf5a6b8da14708a0e7771c5b7a281665cfefa6/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java#L99)):*
<details><summary><i>Previously, tchaurasiya (Tathagat Chaurasiya) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I see exception is still being thrown ?

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java [hunk 9]
```diff
@@ -74,15 +92,257 @@ public List<NGVariable> getOverrideVariables(@NonNull NGServiceOverridesEntity e
...
+      String orgId, String projectId, @NonNull ServiceStepV3Parameters parameters, @NonNull Environment envEntity)
+      throws IOException {
+    if (ParameterField.isNull(parameters.getEnvRef()) || isEmpty(parameters.getEnvRef().getValue())
+        || ParameterField.isNull(parameters.getServiceRef()) || isEmpty(parameters.getServiceRef().getValue())) {
+      throw new InvalidRequestException("Environment Ref or Service Ref given for overrides has not been resolved");
     }
-    if (Boolean.TRUE.equals(entity.getIsV2())) {
-      return mergeOverrideV2Inputs(entity, serviceOverrideInputs);
+    if (ParameterField.isNull(parameters.getInfraId()) || isEmpty(parameters.getInfraId().getValue())) {
+      log.warn("Infra Identifier has not been resolved to get service overrides");
```
Comment (yogesh-chauhan) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java:103]: will be below code be able to handle infraID not being present ?

**Comment by tchaurasiya** [comment]
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java` line 103 at r3](https://reviewable.io/reviews/harness/harness-core/48407#gh-1208386486:-NWaIe0L7TuCD6WJemJx:b-pvxo0w) ([raw file](https://github.com/harness/harness-core/blob/5f9fff0d3e4f509f1c2300d0bc240edca289d059/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java#L103)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

will be below code be able to handle infraID not being present ?
</blockquote></details>

Yes. Discussed offline.

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48407)* status: 52 file(s) to review. 1 of 2 approvals obtained.

**Comment by tchaurasiya** [comment]
trigger uts

**Review by sourabh-awashti — APPROVED** [review]

**Comment by tchaurasiya** [comment]
trigger review

**Pull request merged**